Copy latest code from private to public repository

gshiroma opened the pull request:
This PR updates the GitHub PROTEUS repository from the private DSWx-Optical repository after updates on top of the interface (IF) delivery.

hfattahi reviewed (commented):
Congrats on the first PR @gshiroma in this repository and in OPERA open source. Very impressive work in a short time. Very clean implementation of the algorithm. My comments are very minor regarding better documentation of the code. 

1. There are few argparse arguments with empty docs. The docstring of the functions either missing or not following a specific convention. It will be great to settle on a convention so we all will follow. 

2. I did not see unit tests. If we want to ignore unit tests in this PR, that is ok with me. But please open an issue and preferably suggest the functions that need unit test so that in near future we will add them. 
3. dswx_hls.py turned to a very long script. It would be nice to at least separate the I/O part from the rest of the code which does actual dswx implementation.

vbrancat commented on schemas/dswx_hls.yaml:
```diff
@@ -0,0 +1,35 @@
+            sas_output_file: str()
+       
+        primary_executable:
+            product_type: enum('DSWX_HLS')
```
In my opinion, the schema should have a processing section hosting all the parameters that the user would like to tune.

vbrancat commented on setup.py:
```diff
@@ -0,0 +1,67 @@
+    scripts=['bin/dswx_hls.py'],
+    install_requires=['argparse', 'numpy', 'yamale', 'ruamel',
+                      'osgeo', 'scipy'],
+    # 'tempfile' 'os' 'sys' 'glob' 'mimetypes'
```
If not needed, remove these dependencies in the comment

vbrancat reviewed (commented):
@gshiroma congratulations for having inaugurated the OPERA ADT implementation on public github. Excellent work :D. I left some nit-picking comments. I guess all my comments can be summarized in:
1. Code documentation: let's choose a doc string style and abide to it. Also, more comments on what block of code lines do are very appreciated
2. The runconfig lacks a processing section. We would like to have the user tuning parameters and experiment with them if possible
3. Logging. Is there any requirement on that in terms of e.g., formatting?

yunjunz commented on bin/dswx_hls.py:
```diff
@@ -0,0 +1,1529 @@
+    return image < threshold
+
+def _compute_mask_and_filter_interpreted_layer(interpreted_dswx_band, qa_band):
+    '''
+    QA band - Landsat 8
+    BITS:
+    0 - Cirrus (reserved but not used)
+    1 - Cloud (*)
+    2 - Adjacent to cloud/shadow
+    3 - Cloud shadow (*)
+    4 - Snow/ice (*)
+    5 - Water
+    6-7 - Aerosol quality:
+          00 - Climatology aerosol
+          01 - Low aerosol
+          10 - Moderate aerosol
+          11 - High aerosol
+
+    (*) set output as 9
```
I like the numpy style.

gshiroma replied:
This was painful! But I added the Numpy docstring convention throughout

yunjunz reviewed (commented):
I would suggest a few general recommendations:
1. It would be nice to have a Python module, e.g. named `proteus`, by creating an `src/proteus` directory. We could have the generic file I/O and algo implementations there. Then the recipe script can be much shorter. 
2. Related to the above. `__inti__.py` is only needed for Python to recognize a module, we don't need `__init__.py` and `bin/__init__.py`.

Regarding the doc/code style, besides the nice references that Heresh pointed out. We could also rely on Codacy for general suggestions. @hfattahi could you enable this, or give me permission in the org, and I could do it?

yunjunz commented:
Although really great comments, the sheer number of comments seems overwhelming to me ;). I don't know how we, I mean @gshiroma ;), could handle them all in one PR. To me, a first working version is more important than a non-working version; although it could be further improved, it still deserves a place in the commit history. So maybe take care of the small changes, and leave the big ones (doc style, logging, unit tests, etc.) for another PR or issues?

LiangJYu reviewed (approved):
Aside from a few nits, LGTM.

gshiroma commented on bin/dswx_hls.py:
```diff
@@ -0,0 +1,1529 @@
+percent_slope_moderate = np.nan  # Slope tolerance for moderate confidence water
+percent_slope_wetland = np.nan  # Slope confidence for potential wetland
+percent_slope_low = np.nan  # Slope tolerance for low confidence water or wetland
+hillshade = np.nan  # Hillshade tolerance value
```
Good point, I moved them to `modules/dswx_hls_constants.py`.

gshiroma replied:
Actually, I just changed it again. I moved the default values to the default runconfig.

gshiroma commented on bin/dswx_hls.py:
```diff
@@ -0,0 +1,1529 @@
+    try:
+        from extern.validate_cloud_optimized_geotiff import main as validate_cog
+    except ModuleNotFoundError:
+        logger.info('ERROR could not import module validate_cloud_optimized_geotiff')
```
This was supposed to be a WARNING message. We don't want to stop the execution if the COG (cloud-optimized GeoTIFF) validation script is not available.

gshiroma commented on bin/dswx_hls.py:
```diff
@@ -0,0 +1,1529 @@
+
+    _makedirs(output_file)
+    shape = input_array.shape
+    driver = gdal.GetDriverByName("GTiff")
```
Yes, the outputs are GeoTIFFs

gshiroma commented on modules/dswx_hls.py:
```diff
@@ -0,0 +1,151 @@
+    # create color table
+    mask_ctable = gdal.ColorTable()
+
+    '''
+    set color for each value
+    - Mask cloud shadow bit (0)
+    - Mask snow/ice bit (1)
+    - Mask cloud bit (2)
+    '''
```
Yes, the cloud/cloud-shadow mask is bit encoded.

gshiroma commented on modules/dswx_hls.py:
```diff
@@ -0,0 +1,151 @@
+    dswx_ctable.SetColorEntry(4, (0, 255, 255))  # Cyan - Low confidence 
+                                                 # water or wetland
+    dswx_ctable.SetColorEntry(9, (128, 128, 128))  # Gray - QA masked
+    dswx_ctable.SetColorEntry(255, (0, 0, 0, 255))  # Black - Fill value
```
Good catch! Alpha was supposed to be `0` for these entries.

gshiroma commented on bin/dswx_hls.py:
```diff
@@ -0,0 +1,1529 @@
+    old_settings = np.seterr(invalid='ignore', divide='ignore')
+
+    # Modified Normalized Difference Wetness Index (MNDWI)
+    mndwi = (green - swir1)/(green + swir1)
```
That's a good question. We can confirm this with John, but the rare cases in which `green + swir1` are zero, `mndwi` becomes `inf`. The result of the three tests involving `mndwi` (i.e. tests 1, 4, and 5) will have the same outcome as if `green + swir1` was very low. Same thing for the `ndvi` variable that also has a denominator that can go to zero. The NumPy error messages are silenced [here](https://github.com/opera-adt/PROTEUS/blob/a745431d54a303e909967eacd01e0f409ac5b87a/modules/dswx_hls.py#L321).

hfattahi replied:
So does it mean that those pixels will be simply flagged as "not water" or "invalid data"?

hfattahi replied:
based on offline discussion, it seems that the devision by zero is handled by defaults and results in a very large value instead of inf.

gshiroma commented on bin/dswx_hls.py:
```diff
@@ -0,0 +1,1529 @@
+logger = logging.getLogger('dswx_hls')
+
+# Thresholds
+wigt = 0.0124  # Modified Normalized Difference Wetness Index (MNDWI) Threshold
```
I moved them to the default runconfig. Thanks!

gshiroma commented:
> 

Thank you, @yunjunz ! I created the Python module and named it `modules`. I also removed `bin/__init__.py` following your suggestion. Thanks!

yunjunz commented:
Hi @gshiroma, the python module name should be more meaningful and unique, such as `isce` or `compass`. It is meant to be used in Python as `import compass` for example.

On the repo structure, I would recommend using https://github.com/opera-adt/sentinel1-reader as the reference: everything that the python module depended on should be inside one top-level folder, this includes the `defaults` and `schemas`. I would imagine a structure as:

```bash
PROTEUS
  /bins               #all scripts inside will be copied to "miniconda3/envs/opera/bin" folder after installation via pip/conda
  /docs              #for all non-jupyter-notebook documentations
  /src/proteus   #this folder will be copied to "miniconda3/envs/opera/lib/python3.8/site-packages" after installation via pip/conda
      /defaults
      /schemas
  /tests              #for unit tests
```

gshiroma commented:
> Hi @gshiroma, the python module name should be more meaningful and unique, such as `isce` or `compass`. It is meant to be used in Python as `import compass` for example.
> 
> On the repo structure, I would recommend using https://github.com/opera-adt/sentinel1-reader as the reference: everything that the python module depended on should be inside one top-level folder, this includes the `defaults` and `schemas`. I would imagine a structure as:
> 
> ```shell
> PROTEUS
>   /bins               #all scripts inside will be copied to "miniconda3/envs/opera/bin" folder after installation via pip/conda
>   /docs              #for all non-jupyter-notebook documentations
>   /src/proteus   #this folder will be copied to "miniconda3/envs/opera/lib/python3.8/site-packages" after installation via pip/conda
>       /defaults
>       /schemas
>   /tests              #for unit tests
> ```

Thank you, @yunjunz ! I updated the code to follow your suggested directory structure. I had also to recreate `bin/__init__.py` because the command ` python3 setup.py sdist` was complaining that the file didn't exist. I hope I didn't do anything wrong.

yunjunz commented:
Hi @gshiroma, I committed a few changes to the code structure, as described above. There are a few more things to change:

1. We should use `setuptools` for code installation because `distutils` is deprecated and planned to be removed in python 3.12 (https://docs.python.org/3/library/distutils.html). For `setuptools` approach, only the following 3 files are needed to setup the installation:

```bash
MANIFEST.in
pyproject.toml
setup.cfg or setup.py
```

2. Please `git rebase` this PR against the latest `main` branch.

gshiroma commented:
Thank you for the updates, @yunjunz ! Changes look great!

gshiroma commented:
Hi all , I addressed all review comments that we intended to address on this initial release. Hopefully, we can merge it soon so I can move on to the next updates. Please, let me know if you have any other concerns or comments. Thank you.

yunjunz reviewed (approved):
Looks good to me for merge! Thank you @gshiroma for the patience and addressing all this comments!

vbrancat reviewed (approved):
LGTM. Thanks for having been super patient and have addressed my comments :)

hfattahi commented on src/proteus/defaults/dswx_hls.yaml:
```diff
@@ -0,0 +1,70 @@
+        primary_executable:
+            product_type: DSWX_HLS
+
+        hls_thresholds:
+
+            # Modified Normalized Difference Wetness Index (MNDWI) Threshold
+            wigt: 0.0124
+
+            # Automated Water Extent Shadow Threshold
+            awgt: 0.0
+
+            # Partial Surface Water Test-1 MNDWI Threshold
+            pswt_1_mndwi: -0.44
+
```
I think exposing these parameters as well as the following few thresholds to the user does not make sense. These are very critical parameters that can completely change the algorithm performance and have been determined in results of extensive testing/calibrations. I just don't see a case where a user would want to change these values. I think we should just store them internally  as constants.

hfattahi replied:
Ok as we talked offline, we can keep this implementation as is because it could help with calibration. We can revise this decision after calibration.

hfattahi commented:
Thank you @gshiroma for addressing all the comments. This PR is close to be merged. I put in one comment to move back the thresholds to the code instead of exposing to users.

hfattahi reviewed (approved):
@gshiroma thank you for this interesting work. Thanks again for addressing all reviewers comments. This looks great to me.
